### ISSUE TITLE
OOM error with NUTS on CUDA 

### ISSUE THREAD

**Issue by neerajprad** [pyro-ppl/pyro#1787]
### Issue Description

Issue mentioned in our [forum](https://forum.pyro.ai/t/reducing-memory-requirement-for-nuts/850):

When performing parameter inference with NUTS on CUDA I observed a linear increase in allocated memory that leads to the termination of the processes within a few iterations. This only happens after warmup is finished, and apparently also happens on CPU. The minimum program to reproduce this behavior is:

```
import torch
import pyro
from pyro.infer.mcmc import MCMC, NUTS
import pyro.distributions as dist

data = torch.ones((1000, 1000))
def model():
    x = pyro.sample('x', dist.Normal(0., 1.))
    pyro.sample('data', dist.Normal(x*data, 0.01), obs = data)
nuts_kernel = NUTS(model)
posterior = MCMC(nuts_kernel, 1000, warmup_steps = 0).run()
```

I suspect that the data is somehow copied and stored at each iteration. However, `del posterior` does not free any significant amount of memory (and it appear like only a reference and not copies to the original data object are stored). How can I avoid this excessive memory consumption?

### Environment

(torch 1.01, pyro 0.3.1, python 3.6.7)

**Comment by neerajprad** [comment]
Thanks for the clear report! The memory would increase per iteration since we are storing `num_samples` execution traces, which could be more problematic on the GPU because of higher memory limitation. I think the only way around this would be to store the bare minimal information from the trace as mentioned in this [issue](https://github.com/pyro-ppl/pyro/issues/1725) (yet again!), so I'll definitely make sure that we address this as part of the refactoring.

**Comment by cweniger** [comment]
I used poutine.block to remove the data nodes in `_get_trace` in the HMC and NUTS sampler. This solved the problem for now. See https://github.com/cweniger/pyro for details if you are interested.

**Comment by neerajprad** [comment]
> I used poutine.block to remove the data nodes in _get_trace in the HMC and NUTS sampler. 

That's great! You can certainly use block to not record certain sites in your model. Were you cloning the data in your model? Because otherwise the data nodes should point to the same memory reference and shouldn't consume memory that scales with the number of samples (unless we have some bug in our code!).

**Comment by cweniger** [comment]
I did not clone the data, and indeed the data nodes were all pointing to the same memory reference (inplace modifications modified all at once). Still, blocking these nodes solved the problem. I'm also confused by what is going on, but for now it seems to work.

**Comment by fehiepsi** [comment]
I think that `data` node’s fn will hold different loc/scale for each trace, so we get OOM. This can be another reason for not holding the whole trace in MCMC. I believe #1730 will also be resolved when we refactor out MCMC.
